index.js

```diff
@@ -2120,24 +2124,41 @@
   //. Cons('red', Cons('blue', Cons('green', Nil)))
   //. ```
   function sortBy(f, foldable) {
-    var rs = reduce(function(xs, x) {
-      var fx = f(x);
-      var lower = 0;
-      var upper = xs.length;
-      while (lower < upper) {
-        var idx = Math.floor((lower + upper) / 2);
-        if (lte(xs[idx].fx, fx)) lower = idx + 1; else upper = idx;
-      }
-      xs.splice(lower, 0, {x: x, fx: fx});
+    var foldableIsArray = Array.isArray(foldable);
+    var idx = 0;
+    function reducer(xs, x) {
+      xs.push({idx: idx, x: x, prop: f(x)});
+      idx += 1;
       return xs;
-    }, [], foldable);
-
-    var F = foldable.constructor;
-    var result = empty(F);
-    for (var idx = 0; idx < rs.length; idx += 1) {
-      result = concat(result, of(F, rs[idx].x));
     }
-    return result;
+    var rs = reduce(reducer, [], foldable);
+    if (rs.length === 0) {
+      return foldable;
+    }
+    var lteFn =
+      typeof rs[0].prop === 'string' || typeof rs[0].prop === 'number' ?
+        function(a, b) { return a <= b; } :
```

Comment:
This function can be hoisted. May slightly benefit performance.

Reply:
The fast-path for numbers and strings is very important for performance.

I think we should consider whether it's worth the code complexity and performance to handle `NaN`. For instance, `Array$prototype$lte` doesn't handle sparse arrays because sparse arrays can be considered to not be a part of the array type we define. In the same way `Number$prototype$lte` could not handle `NaN` as `NaN` could be considered not part of the number type we define.

Reply:
I think it would be best for these changes to be refactoring only. In a subsequent pull request you could make the case for not supporting `NaN` by showing the performance difference.

Reply:
:zap: